Add Mochawesome Report Generator for Cypress Tests

jessica-lau opened the pull request:
### What does this pull request aim to achieve? 💯

Add mochawesome report generator for Cypress tests. This creates a full fledged HTML report that helps visualize your test suites. Here is an example to what the output result looks like:

![image](https://user-images.githubusercontent.com/56090896/97114086-c3d74c00-16c4-11eb-8bfa-2164c6fe1713.png)


<!--_Please do **NOT** submit this PR if you have not done the following:_-->

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

jessica-lau committed:
added mochawsome report generator for cypresse2e tests

jessica-lau commented:
@JayashreeSamantasinghar @TimeaKovacs86 @galeza please see readme file on how to generate the report if you'd like to try it out yourself. Thank you!

galeza commented:
hi @jessica-lau 

So current flow is as follow ? :
1. Start tests (I'm using docker-compose)
2. When tests finish
3. cd cypresse2e and run npm run generate:html:report

When I run npm run generate:html:report I got error:
```
> cypresse2e@1.0.0 merge:reports /Users/agagaleza/Documents/work/FightPandemics/cypresse2e
> mochawesome-merge mochawesome-report/*.json > cypress-tests-report.json

sh: mochawesome-merge: command not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! cypresse2e@1.0.0 merge:reports: `mochawesome-merge mochawesome-report/*.json > cypress-tests-report.json`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the cypresse2e@1.0.0 merge:reports script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/agagaleza/.npm/_logs/2020-10-26T13_28_05_176Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
```

jessica-lau commented:
> hi @jessica-lau
> 
> So current flow is as follow ? :
> 
> 1. Start tests (I'm using docker-compose)
> 2. When tests finish
> 3. cd cypresse2e and run npm run generate:html:report
> 
> When I run npm run generate:html:report I got error:
> 
> ```
> > cypresse2e@1.0.0 merge:reports /Users/agagaleza/Documents/work/FightPandemics/cypresse2e
> > mochawesome-merge mochawesome-report/*.json > cypress-tests-report.json
> 
> sh: mochawesome-merge: command not found
> npm ERR! code ELIFECYCLE
> npm ERR! syscall spawn
> npm ERR! file sh
> npm ERR! errno ENOENT
> npm ERR! cypresse2e@1.0.0 merge:reports: `mochawesome-merge mochawesome-report/*.json > cypress-tests-report.json`
> npm ERR! spawn ENOENT
> npm ERR! 
> npm ERR! Failed at the cypresse2e@1.0.0 merge:reports script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/agagaleza/.npm/_logs/2020-10-26T13_28_05_176Z-debug.log
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> ```

hey @galeza, did you run 'npm install' in cypresse2e directory before doing the steps mentioned above? I was running it locally on my computer, sorry for the confusion. Will make the changes in the Readme file.

So try this:
1. Run 'npm install' in cd cypresse2e
2. Run 'docker-compose run cypress-e2e'
3. After tests are done run 'npm run generate:html:report'

Let me know if this works.

adorapham commented:
@jessica-lau and All , I am able to run and see it works. please see the attached screenshots.  It looks very good.  For npm run generate:html:report, I run under the power shell.
![image](https://user-images.githubusercontent.com/70918430/97205508-cf824b80-1774-11eb-8c80-17e1f27ab847.png)
![image](https://user-images.githubusercontent.com/70918430/97205738-140de700-1775-11eb-8a60-2eac150ebf34.png)
My comments/questions 
1-it is nice if we have the env information (URL)
2-if we want to run only one spec file, which command should we use, currently we use npm run test.

galeza commented:
Hi @jessica-lau @adorapham 

I run: npm i --D mocha mochawesome mochawesome-merge mochawesome-report-generator
and I was able to generate a report.

The problem is that we need to be able to generate report in the docker-compose. 
When we run tests against a PR/code we should get a report. 
Tomorrow I will try to add: `I run: npm i --D mocha mochawesome mochawesome-merge mochawesome-report-generator` in the dockerfile and then in the docker-compose mount volumes like here: ` - ./cypresse2e/cypress:/app/cypress`
We should be able to run it with npx command like here: https://www.linkedin.com/pulse/containerised-user-journey-automation-using-cypress-karishma-sharma/

TimeaKovacs86 reviewed (commented):
Hi @jessica-lau ! I tried to run the report generation script, and got the same result as @galeza, as in I had nothing in the cypress-tests-report.json so it failed to merge reports.

jessica-lau commented:
> Hi @jessica-lau @adorapham
> 
> I run: npm i --D mocha mochawesome mochawesome-merge mochawesome-report-generator
> and I was able to generate a report.
> 
> The problem is that we need to be able to generate report in the docker-compose.
> When we run tests against a PR/code we should get a report.
> Tomorrow I will try to add: `I run: npm i --D mocha mochawesome mochawesome-merge mochawesome-report-generator` in the dockerfile and then in the docker-compose mount volumes like here: ` - ./cypresse2e/cypress:/app/cypress`
> We should be able to run it with npx command like here: https://www.linkedin.com/pulse/containerised-user-journey-automation-using-cypress-karishma-sharma/

@galeza agreed that's the problem, let me take a deeper look into this and see if I can fix it. Will update everyone on it. Thanks!

jessica-lau commented:
@galeza if a test fails when running all tests, it gives an 'exit code 6' and stops the docker-compose run and doesn't generate an html report (this is a default with mochawesome). 

Should we automatically generate an html report by setting 'exit code 0' even if there's a fail test? Or should we generate the html report separately in docker-compose command?

jessica-lau committed:
revised ReadMe.md and updated scripts and docker-compose.yml

galeza commented:
> @galeza if a test fails when running all tests, it gives an 'exit code 6' and stops the docker-compose run and doesn't generate an html report (this is a default with mochawesome).
> 
> Should we automatically generate an html report by setting 'exit code 0' even if there's a fail test? Or should we generate the html report separately in docker-compose command?

Hi @jessica-lau 

When I run `docker-compose run cypress-e2e` it runs only one spec, the last one. Do you see the same behavior? :
```
> cypresse2e@1.0.0 test /app
> cypress run --spec cypress/integration/e2eTests/post/post.spec.js
```
![Screen Shot 2020-10-27 at 12 03 34](https://user-images.githubusercontent.com/21987083/97293118-8271a880-184c-11eb-87ba-f7bc12779fd7.png)

When I removed all containers and re-downloaded them `i have a following issue:
![Screen Shot 2020-10-27 at 12 05 51](https://user-images.githubusercontent.com/21987083/97293286-c5338080-184c-11eb-8032-bf39b90d6366.png)
I think that two services with the same name are going to be created and it causes a problem.

In one of my tests I saw a report generated when all tests failed:
![Screen Shot 2020-10-27 at 12 08 44](https://user-images.githubusercontent.com/21987083/97293602-33784300-184d-11eb-9855-b7f4f419bbc5.png)
Did you ask about this situation or maybe when eg. client is not compiled and all specs fail?

I was thinking about asking devops if we go the right way? What you think about it?
The final goal is to add new github action workflow, run tests and get a report. Probably we will use a command `docker-compose up --exit-code-from cypress-e2e` to run tests and get the exit code. Then report should be available. Something like here (example from article: https://docs.github.com/en/free-pro-team@latest/actions/guides/storing-workflow-data-as-artifacts):
![Screen Shot 2020-10-27 at 12 13 05](https://user-images.githubusercontent.com/21987083/97294083-d4ff9480-184d-11eb-9d80-83f2689c35e5.png)

adorapham commented:
@jessica-lau @galeza , I follow the instruction running both docker-compose and local - both see only one e2eTests/post/post.spec.js  running
**docker-compose** don't see the 'mochawesome html' report generate automatically. only see one mochawesome JSON file generated.  I think we forget to include the command to generate the report npm run report
![image](https://user-images.githubusercontent.com/70918430/97354626-eac38880-1852-11eb-8424-cbc6cbedbffa.png)

**local**  npm run test
![image](https://user-images.githubusercontent.com/70918430/97354680-02027600-1853-11eb-840a-c2e2103570a0.png)
 npm run report
![image](https://user-images.githubusercontent.com/70918430/97354934-6a515780-1853-11eb-8500-e1400f4f0f19.png)

in package.json  , i see only 

        "test": "cypress run --spec cypress/integration/e2eTests/post/post.spec.js",

jessica-lau committed:
updated readme.md and mochawesome scripts

jessica-lau commented:
hey @galeza @adorapham the reason you only saw one test execute is because I was trying out a couple of things with a single spec the other day then forgot to change it back to include all tests. Anyway, I fixed this and you should see all the tests run now. I also finished updating the readme.md. Sorry about that. Make sure to run `docker-compose build cypress-e2e-tests` if you're running on docker as I made some changes. 

@adorapham the command to run the html report in docker-compose is `docker-compose run cypress-e2e-report`. I added this into readme.

@galeza I changed the container_name in docker-compose file so that they have different names now but it's still having issues when running `docker-compose up --exit-code-from cypress-e2e-tests` and no mochawesome json generates. Are you seeing this error when you run it again with this command?

This is what I'm seeing:
![image](https://user-images.githubusercontent.com/56090896/97376960-25e6ab80-1895-11eb-930e-db535e906dd2.png)

Yea, I think we should ask devops for some guidance.

For now, maybe we should keep the process of running the tests and generating the report as two separate commands like how I stated in the readme (in order to get the html report even if a test fails), until this exit code situation gets figured out? 

Let me know if I'm missing anything or if you see any other issues.

adorapham commented:
@jessica-lau , @galeza  when we run docker-compose down and then docker-compose up , we see all the json files generated but the report command somehow is unable to reach. **It seems we are almost there**.  I agree **Yea, I think we should ask devops for some guidance** for future improvement.
 When i move to cypresse2e and run all docker-compose build and run, i see the report for all tested are generated.  

![image](https://user-images.githubusercontent.com/70918430/97487376-08a7f080-191a-11eb-8dfc-2ec11c3427ac.png)

![image](https://user-images.githubusercontent.com/70918430/97487183-bf57a100-1919-11eb-9a3c-b8448cd9ad0a.png)

**For local , npm run test has issue so i use different command** 


 
![image](https://user-images.githubusercontent.com/70918430/97486237-63d8e380-1918-11eb-940e-a9c2f5deeb79.png)

i re-run in local with npm run test conmmand in bash and it works

![image](https://user-images.githubusercontent.com/70918430/97576953-01c9be00-19ac-11eb-92f5-8bc24ed3f994.png)


LGTM  , Thank you @jessica-lau

galeza commented:
Hi @jessica-lau 
I removed containers and rebuild them.
I ran :
```
docker-compose build cypress-e2e-tests
docker-compose run cypress-e2e-tests
docker-compose run cypress-e2e-report
```
Report is generated. I only see errors when tests fails:
![Screen Shot 2020-10-29 at 14 40 58](https://user-images.githubusercontent.com/21987083/97581408-d1f2d880-19f4-11eb-8802-68d3190a8ca0.png)

BTW: I think that two tests are flaky:
![Screen Shot 2020-10-28 at 16 45 41](https://user-images.githubusercontent.com/21987083/97581469-e59e3f00-19f4-11eb-9422-c2d211d9380d.png)
![Screen Shot 2020-10-28 at 16 45 30](https://user-images.githubusercontent.com/21987083/97581477-e6cf6c00-19f4-11eb-82a5-ead2d203f99a.png)
Do you think that we could temporarily disable them by using xit?

galeza commented:
Hi @mannykary 

We have a question about generating artefacts in github action.
Currently we have screenshots and report generated. We would like to run cypress tests against each PR so we need to add github action workflow.
Would it be possible to store screenshots and report as artefacts like in the picture below:
![Screen Shot 2020-10-27 at 12 13 05](https://user-images.githubusercontent.com/21987083/97583860-b2a97a80-19f7-11eb-8e0b-086e65c9d010.png)

In this PR we run:
```
cd cypresse2e
docker-compose build cypress-e2e-tests
docker-compose run cypress-e2e-tests
docker-compose run cypress-e2e-report
```
Artefacts are stored in the  `./cypresse2e/mochawesome-report/:/app/mochawesome-report/` and  `./cypresse2e/cypress:/app/cypress` folders.
Are we going in the right direction?

jessica-lau committed:
Merge remote-tracking branch 'origin' into feature/cypress-add-mochawesome

jessica-lau committed:
append skip to location input test case for offer help and request help spec

jessica-lau commented:
> Hi @jessica-lau
> I removed containers and rebuild them.
> I ran :
> 
> ```
> docker-compose build cypress-e2e-tests
> docker-compose run cypress-e2e-tests
> docker-compose run cypress-e2e-report
> ```
> 
> Report is generated. I only see errors when tests fails:
> ![Screen Shot 2020-10-29 at 14 40 58](https://user-images.githubusercontent.com/21987083/97581408-d1f2d880-19f4-11eb-8802-68d3190a8ca0.png)
> 
> BTW: I think that two tests are flaky:
> ![Screen Shot 2020-10-28 at 16 45 41](https://user-images.githubusercontent.com/21987083/97581469-e59e3f00-19f4-11eb-9422-c2d211d9380d.png)
> ![Screen Shot 2020-10-28 at 16 45 30](https://user-images.githubusercontent.com/21987083/97581477-e6cf6c00-19f4-11eb-82a5-ead2d203f99a.png)
> Do you think that we could temporarily disable them by using xit?

@galeza yea those two location list test cases are pretty flaky - sometimes they run fine and other times they don't. I appended 'skip' to them for now.

haopham288 reviewed (commented):
New changes are coming so i move back to pending review. (from Adora Pham)

Updated: re-run with new commit - both docker and local works . LGTM - move back to approved 
![image](https://user-images.githubusercontent.com/70918430/97901964-ba1b9d00-1cf1-11eb-801e-a9b2dda6753c.png)

mannykary commented:
> Hi @mannykary
> 
> We have a question about generating artefacts in github action.
> Currently we have screenshots and report generated. We would like to run cypress tests against each PR so we need to add github action workflow.
> Would it be possible to store screenshots and report as artefacts like in the picture below:
> ![Screen Shot 2020-10-27 at 12 13 05](https://user-images.githubusercontent.com/21987083/97583860-b2a97a80-19f7-11eb-8e0b-086e65c9d010.png)
> 
> In this PR we run:
> 
> ```
> cd cypresse2e
> docker-compose build cypress-e2e-tests
> docker-compose run cypress-e2e-tests
> docker-compose run cypress-e2e-report
> ```
> 
> Artefacts are stored in the `./cypresse2e/mochawesome-report/:/app/mochawesome-report/` and `./cypresse2e/cypress:/app/cypress` folders.
> Are we going in the right direction?

I did a bit of research and I think this should be possible: https://github.com/cypress-io/github-action#artifacts

As long as you set the volume mounts as you described it should be possible.

galeza commented:
@mannykary 
Thank you!

galeza reviewed (approved):
LGTM

idjevm reviewed (changes requested):
Lots of merge conflicts, please take a look and try to resolve them.

jessica-lau committed:
resolved merge conflicts

jessica-lau commented:
@idjevm I fixed the conflicts, please merge. Thank you! : )

idjevm commented on package-lock.json:
```diff
@@ -2,27 +2,445 @@
   "dependencies": {
-    "@babel/runtime": {
-      "version": "7.11.2",
-      "resolved": "https://registry.npmjs.org/@babel/runtime/-/runtime-7.11.2.tgz",
-      "integrity": "sha512-TeWkU52so0mPtDcaCTxNBI/IHiz0pZgr8VEFqXFtZWpYD08ZB6FaSwVAS8MKRQAP3bYKiVjwysOJgMFY28o6Tw==",
+    "@babel/code-frame": {
```
Sorry @jessica-lau, I actually just noticed that this is from the root folder. 

We probably want to avoid making changes to this one and also within the cypress service, unless new packages are added. Could you take a look and see if you can remove them? I could do it from my end if needed. Let me know

idjevm commented on cypresse2e/package-lock.json:
```diff
@@ -145,23 +145,11 @@
-          "dev": true
-        }
-      }
+    "@ungap/promise-all-settled": {
```
and I mean the whole file, with these two. It would be probably best to change it to only commit changes if there were new packages added.

idjevm reviewed (changes requested):
My apologies, I just noticed that these two files were having huge diffs, so we would probably not want to do that, especially the one from the root folder

jessica-lau committed:
fixed package-lock.json files

jessica-lau commented:
@idjevm sorry about that. I fixed the package-lock.json in the root folder. However, packages related to mochawesome were added so I had to do an npm install for the package-lock.json in the cypress folder.

idjevm reviewed (approved):
Awesome! This looks good @jessica-lau ! Let's merge it after we deploy prod today!

jessica-lau committed:
fixed merge conflicts in selector.js

jessica-lau committed:
fixed description

jessica-lau committed:
fixed description

galeza commented:
@jessica-lau 
Can we merge it or you want to work more on it? Let us know.

idjevm commented:
Looks good now! mergin! Thanks @jessica-lau
